sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -195,17 +207,20 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
           fileFormatClass,
           None)
         val logicalRelation = cached.getOrElse {
+          val fileIndex = new InMemoryFileIndex(sparkSession, Seq(rootPath), options, None)
```

Comment:
I was thinking that ```InMemoryFileIndex``` had similar semantics to ```CatalogFileIndex``` where the leaf files are only listed if  ```listFiles()``` is invoked. I can make the ```fileIndex``` arg of ```inferIfNeeded()``` a ```Option[FileIndex]``` and create the ```InMemoryFileIndex``` within that method instead if we need to infer a schema and no file index is provided.